hypervisor/arch/x86/trusty.c

```diff
@@ -50,6 +50,222 @@ static struct key_info g_key_info = {
+ * @param gpa_rebased gpa rebased to offset xxx (511G_OFFSET)
+ *
+ */
+static void create_secure_world_ept(struct vm *vm, uint64_t gpa_orig,
+		int64_t size, uint64_t gpa_rebased)
+{
+	uint64_t nworld_pml4e = 0;
+	uint64_t sworld_pml4e = 0;
+	struct entry_params entry;
+	struct map_params map_params;
+	uint64_t gpa_uos = gpa_orig;
+	uint64_t gpa_sos;
+	uint64_t adjust_size;
+	uint64_t mod;
+	uint64_t hpa = gpa2hpa(vm, gpa_uos);
+	void *sub_table_addr = NULL, *pml4_base = NULL;
+	int i;
+	struct vm *vm0 = get_vm_from_vmid(0);
+	struct vcpu *vcpu;
+
+	if (vm0 == NULL) {
+		pr_err("Parse vm0 context failed.");
+		return;
+	}
+
+	if (!vm->sworld_control.sworld_enabled
+			|| vm->arch_vm.sworld_eptp != 0UL) {
+		pr_err("Sworld is not enabled or Sworld eptp is not NULL");
+		return;
+	}
+
+	/* Backup secure world info, will be used when
+	 * destroy secure world
+	 */
+	vm->sworld_control.sworld_memory.base_gpa = gpa_rebased;
+	vm->sworld_control.sworld_memory.base_hpa = hpa;
+	vm->sworld_control.sworld_memory.length = size;
+	vm->sworld_control.sworld_memory.gpa_sos = hpa2gpa(vm0, hpa);
+	vm->sworld_control.sworld_memory.gpa_uos = gpa_orig;
+
+	/* Copy PDPT entries from Normal world to Secure world
+	 * Secure world can access Normal World's memory,
+	 * but Normal World can not access Secure World's memory.
+	 * The PML4/PDPT for Secure world are separated from
+	 * Normal World.PD/PT are shared in both Secure world's EPT
+	 * and Normal World's EPT
+	 */
+	pml4_base = alloc_paging_struct();
+	vm->arch_vm.sworld_eptp = HVA2HPA(pml4_base);
+
+	/* The trusty memory is remapped to guest physical address
+	 * of gpa_rebased to gpa_rebased + size
+	 */
+	sub_table_addr = alloc_paging_struct();
+	sworld_pml4e = HVA2HPA(sub_table_addr) | IA32E_EPT_R_BIT |
+				IA32E_EPT_W_BIT |
+				IA32E_EPT_X_BIT;
+	mem_write64(pml4_base, sworld_pml4e);
+
+	nworld_pml4e = mem_read64(HPA2HVA(vm->arch_vm.nworld_eptp));
+	(void)memcpy_s(HPA2HVA(sworld_pml4e & IA32E_REF_MASK), CPU_PAGE_SIZE,
+			HPA2HVA(nworld_pml4e & IA32E_REF_MASK), CPU_PAGE_SIZE);
+
+	/* Unmap gpa_orig~gpa_orig+size from guest normal world ept mapping */
+	map_params.page_table_type = PTT_EPT;
+
+	while (size > 0) {
+		map_params.pml4_base = HPA2HVA(vm->arch_vm.nworld_eptp);
+		map_params.pml4_inverted = HPA2HVA(vm->arch_vm.m2p);
+		obtain_last_page_table_entry(&map_params, &entry,
+				(void *)gpa_uos, true);
+		mod = (gpa_uos % entry.page_size);
+		adjust_size = (mod) ? (entry.page_size - mod) : entry.page_size;
+		if ((uint64_t)size < entry.page_size)
+			adjust_size = size;
+		hpa = gpa2hpa(vm, gpa_uos);
+
+		/* Unmap from normal world */
+		unmap_mem(&map_params, (void *)hpa,
+			(void *)gpa_uos, adjust_size, 0);
+
+		/* Map to secure world */
+		map_params.pml4_base = HPA2HVA(vm->arch_vm.sworld_eptp);
+		map_mem(&map_params, (void *)hpa,
+			(void *)gpa_rebased, adjust_size,
+			(IA32E_EPT_R_BIT |
+			 IA32E_EPT_W_BIT |
+			 IA32E_EPT_X_BIT |
+			 IA32E_EPT_WB));
```

Comment:
IA32E_EPT_xxx are uint64_t, while map_mem expects a uint32_t flags for now. @lifeix will the flags be updated to uint64_t?